README.rst
aiohttp_jrpc
============
.. image:: https://travis-ci.org/zloidemon/aiohttp_jrpc.svg?branch=master
    :target: https://travis-ci.org/zloidemon/aiohttp_jrpc
.. image:: https://coveralls.io/repos/zloidemon/aiohttp_jrpc/badge.svg
    :target: https://coveralls.io/r/zloidemon/aiohttp_jrpc
.. image:: https://badge.fury.io/py/aiohttp_jrpc.svg
    :target: https://badge.fury.io/py/aiohttp_jrpc

jsonrpc_ protocol implementation for `aiohttp.web`__.

__ aiohttp_web_


Example server
--------------

.. code:: python

    import asyncio
    from aiohttp import web
    from aiohttp_jrpc import Service, JError, jrpc_errorhandler_middleware

    SCH = {
        "type": "object",
        "properties": {
            "data": {"type": "string"},
        },
    }

    @asyncio.coroutine
    def custom_errorhandler_middleware(app, handler):
        @asyncio.coroutine
        def middleware(request):
            try:
                return (yield from handler(request))
            except Exception:
                """ Custom errors: -32000 to -32099 """
                return JError().custom(-32000, "Example error")
        return middleware

    class MyJRPC(Service):
        @Service.valid(SCH)
        def hello(self, ctx, data):
            if data["data"] == "hello":
                return {"status": "hi"}
            return {"status": data}

        def error(self, ctx, data):
            raise Exception("Error which will catch middleware")

        def no_valid(self, ctx, data):
            """ Method without validation incommig data """
            return {"status": "ok"}

    @asyncio.coroutine
    def init(loop):
        app = web.Application(loop=loop, middlewares=[jrpc_errorhandler_middleware])
        #app = web.Application(loop=loop, middlewares=[custom_errorhandler_middleware])
        app.router.add_route('POST', "/api", MyJRPC)

        srv = yield from loop.create_server(app.make_handler(),
                                            "127.0.0.1", 8080)
        print("Server started at http://127.0.0.1:8080")
        return srv

    loop = asyncio.get_event_loop()
    loop.run_until_complete(init(loop))
    try:
        loop.run_forever()
    except KeyboardInterrupt:
        pass

Example client
--------------

.. code:: python

    import asyncio
    import aiohttp
    from aiohttp_jrpc import Client,InvalidResponse

    Remote = Client('http://localhost:8080/api')

    @asyncio.coroutine
    def rpc_call():
        try:
            rsp = yield from Remote.request('hello', {'data': 'hello'})
            return rsp
        except InvalidResponse as err:
            return err
        except Exception as err:
            return err
        return False

    loop = asyncio.get_event_loop()
    content = loop.run_until_complete(rpc_call())
    print(content.result)
    loop.close()

License
-------

``aiohttp_jrpc`` BSD license.


.. _jsonrpc: http://www.jsonrpc.org/specification
.. _aiohttp_web: http://aiohttp.readthedocs.org/en/latest/web.html


CHANGES.txt
CHANGES
=======

0.1.0 (2016-02-20)

* Added client and tests
* Changed BSD v3 to BSD v2 license

0.0.3 (2015-10-27)

* Fix messages of protocol errors
* Fix tests and add tests for custom errors
* Fix example bugs
* Added custom middleware to example for handle errors

0.0.2 (2015-10-22)

* Added middleware to catch exceptions
* Testing internal error

0.0.1 (2015-10-18)

* Init release


tests/utils.py
import asyncio
from aiohttp_jrpc import Service, JError


PARSE_ERROR = {
    'jsonrpc': '2.0', 'id': None,
    'error': {'code': -32700, 'message': 'Parse error'}
}
INVALID_REQUEST = {
    'jsonrpc': '2.0', 'id': None,
    'error': {'code': -32600, 'message': 'Invalid Request'}
}
NOT_FOUND = {
    'jsonrpc': '2.0', 'id': None,
    'error': {'code': -32601, 'message': 'Method not found'}
}
INVALID_PARAMS = {
    'jsonrpc': '2.0', 'id': None,
    'error': {'code': -32602, 'message': 'Invalid params'}
}
INTERNAL_ERROR = {
    'jsonrpc': '2.0', 'id': None,
    'error': {'code': -32603, 'message': 'Internal error'}
}
CUSTOM_ERROR_GT = {
    'jsonrpc': '2.0', 'id': None,
    'error': {'code': -32000, 'message': 'Custom error gt'}
}
CUSTOM_ERROR_LT = {
    'jsonrpc': '2.0', 'id': None,
    'error': {'code': -32099, 'message': 'Custom error lt'}
}

REQ_SCHEM = {
    "type": "object",
    "properties": {
        "data": {"type": "string"},
    }
}


@asyncio.coroutine
def custom_errorhandler_middleware(app, handler):
    @asyncio.coroutine
    def middleware(request):
        try:
            return (yield from handler(request))
        except AttributeError:
            return JError().custom(-32000, 'Custom error gt')
        except LookupError:
            return JError().custom(-32099, 'Custom error lt')
        except NameError:
            return JError().custom(-32100, 'Bad custom error')
        except Exception:
            return JError().custom(-31999, 'Bad custom error')
    return middleware


class MyService(Service):
    def hello(self, ctx, data):
        return {"a": "b"}

    @Service.valid(REQ_SCHEM)
    def v_hello(self, ctx, data):
        if data["data"] == "TEST":
            return {"status": "OK"}
        return {"status": "ok"}

    def err_exc(self, ctx, data):
        raise Exception("test middleware, exception is ok")

    def err_exc2(self, ctx, data):
        raise NameError("test custom middleware, exception is ok")

    def err_gt(self, ctx, data):
        raise AttributeError("test custom middleware, exception is ok")

    def err_lt(self, ctx, data):
        raise LookupError("test custom middleware, exception is ok")


def create_response(id=None, result=None):
    return {"jsonrpc": "2.0", "id": id, "result": result}


def create_request(method, id=None, params=None):
    return {
        "jsonrpc": "2.0", "id": id,
        "method": method, "params": params
    }


tests/test_jrpc_client.py
import asyncio
import socket
import unittest
from aiohttp import web
from aiohttp_jrpc import Client, Response
from utils import custom_errorhandler_middleware, MyService
from utils import create_response
from utils import (NOT_FOUND, INVALID_PARAMS, INTERNAL_ERROR,
                   CUSTOM_ERROR_GT, CUSTOM_ERROR_LT)


class TestClient(unittest.TestCase):

    def setUp(self):
        self.loop = asyncio.new_event_loop()
        asyncio.set_event_loop(None)

    def tearDown(self):
        self.loop.run_until_complete(self.handler.finish_connections())
        self.srv.close()
        self.loop.run_until_complete(self.srv.wait_closed())
        self.loop.close()

    def find_unused_port(self):
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.bind(('127.0.0.1', 0))
        port = sock.getsockname()[1]
        sock.close()
        return port

    @asyncio.coroutine
    def request_wrapper(self, request):
        """ It's acctually need for tests on travis I could not reproduce """
        return (yield from MyService(request))

    @asyncio.coroutine
    def create_server(self, middlewares=[]):
        app = web.Application(loop=self.loop,
                              middlewares=middlewares)

        port = self.find_unused_port()
        self.handler = app.make_handler(
            debug=False, keep_alive_on=False)
        app.router.add_route('*', '/', self.request_wrapper)
        srv = yield from self.loop.create_server(
            self.handler, '127.0.0.1', port)
        url = "http://127.0.0.1:{}/".format(port)
        self.srv = srv
        client = Client(url, loop=self.loop)
        return app, srv, client

    def test_validate(self):
        @asyncio.coroutine
        def call(check, method, data=None, id=None):
            app, srv, client = yield from self.create_server(middlewares=[
                custom_errorhandler_middleware])

            resp = Response(**check)

            if not id:
                id = resp.id

            ret = yield from client.call(method, data, id=id)

            self.assertEqual(resp.error, ret.error)
            self.assertEqual(resp.result, ret.result)

            if resp.id:
                self.assertEqual(resp.id, ret.id)

        self.loop.run_until_complete(call(INVALID_PARAMS, "v_hello"))
        self.loop.run_until_complete(
            call(create_response(result={"status": "ok"}),
                 "v_hello", {"data": "ok"}))
        self.loop.run_until_complete(
            call(create_response(1234, {"status": "ok"}),
                 "v_hello", {"data": "ok"}, 1234))
        self.loop.run_until_complete(
            call(create_response(None, {"status": "ok"}),
                 "v_hello", {"data": "ok"}, None))
        self.loop.run_until_complete(
            call(create_response("1", {"status": "OK"}),
                 "v_hello", {"data": "TEST"}, "1"))
        self.loop.run_until_complete(
            call(create_response(True, {"status": "ok"}),
                 "v_hello", {"data": "ok"}, True))
        self.loop.run_until_complete(
            call(create_response(False, {"status": "ok"}),
                 "v_hello", {"data": "ok"}, False))

        self.loop.run_until_complete(call(NOT_FOUND, "not_found"))
        self.loop.run_until_complete(call(INVALID_PARAMS, "v_hello"))
        self.loop.run_until_complete(call(INTERNAL_ERROR, "err_exc"))
        self.loop.run_until_complete(call(INTERNAL_ERROR, "err_exc2"))
        self.loop.run_until_complete(call(CUSTOM_ERROR_GT, "err_gt"))
        self.loop.run_until_complete(call(CUSTOM_ERROR_LT, "err_lt"))


tests/test_jrpc_service.py
import asyncio
import socket
import unittest
import json
import aiohttp
from aiohttp import web
from aiohttp_jrpc import jrpc_errorhandler_middleware
from utils import custom_errorhandler_middleware, MyService
from utils import create_response, create_request
from utils import (PARSE_ERROR, INVALID_REQUEST, NOT_FOUND, INVALID_PARAMS,
                   INTERNAL_ERROR, CUSTOM_ERROR_GT, CUSTOM_ERROR_LT)


class TestService(unittest.TestCase):

    def setUp(self):
        self.loop = asyncio.new_event_loop()
        asyncio.set_event_loop(None)
        self.client = aiohttp.ClientSession(loop=self.loop)

    def tearDown(self):
        self.client.close()
        self.loop.run_until_complete(self.handler.finish_connections())
        self.srv.close()
        self.loop.run_until_complete(self.srv.wait_closed())
        self.loop.close()

    def find_unused_port(self):
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.bind(('127.0.0.1', 0))
        port = sock.getsockname()[1]
        sock.close()
        return port

    @asyncio.coroutine
    def request_wrapper(self, request):
        """ It's acctually need for tests on travis I could not reproduce """
        return (yield from MyService(request))

    @asyncio.coroutine
    def create_server(self, middlewares=[]):
        app = web.Application(loop=self.loop,
                              middlewares=middlewares)

        port = self.find_unused_port()
        self.handler = app.make_handler(
            debug=False, keep_alive_on=False)
        app.router.add_route('*', '/', self.request_wrapper)
        srv = yield from self.loop.create_server(
            self.handler, '127.0.0.1', port)
        url = "http://127.0.0.1:{}/".format(port)
        self.srv = srv
        return app, srv, url

    def test_errors(self):

        @asyncio.coroutine
        def get(check):
            app, srv, url = yield from self.create_server(middlewares=[
                jrpc_errorhandler_middleware])
            resp = yield from self.client.get(url)
            self.assertEqual(200, resp.status)
            self.assertEqual(check, (yield from resp.json()))
            self.assertEqual(None, (yield from resp.release()))

        @asyncio.coroutine
        def post(check, data=None):
            app, srv, url = yield from self.create_server(middlewares=[
                jrpc_errorhandler_middleware])
            resp = yield from self.client.post(url, data=json.dumps(data))
            self.assertEqual(200, resp.status)
            self.assertEqual(check, (yield from resp.json()))
            self.assertEqual(None, (yield from resp.release()))

        self.loop.run_until_complete(get(PARSE_ERROR))
        self.loop.run_until_complete(post(INVALID_REQUEST))
        self.loop.run_until_complete(post(INVALID_REQUEST, {"example": None}))
        self.loop.run_until_complete(post(NOT_FOUND,
                                          create_request("not_found")))
        self.loop.run_until_complete(post(INVALID_PARAMS,
                                          create_request("v_hello")))
        self.loop.run_until_complete(post(INTERNAL_ERROR,
                                          create_request("err_exc")))

    def test_custom_error(self):

        @asyncio.coroutine
        def post(check, data=None):
            app, srv, url = yield from self.create_server(middlewares=[
                custom_errorhandler_middleware])
            resp = yield from self.client.post(url, data=json.dumps(data))
            self.assertEqual(200, resp.status)
            self.assertEqual(check, (yield from resp.json()))
            self.assertEqual(None, (yield from resp.release()))

        self.loop.run_until_complete(post(INTERNAL_ERROR,
                                          create_request("err_exc")))
        self.loop.run_until_complete(post(INTERNAL_ERROR,
                                          create_request("err_exc2")))
        self.loop.run_until_complete(post(CUSTOM_ERROR_GT,
                                          create_request("err_gt")))
        self.loop.run_until_complete(post(CUSTOM_ERROR_LT,
                                          create_request("err_lt")))

    def test_validate(self):
        @asyncio.coroutine
        def post(check, data=None):
            app, srv, url = yield from self.create_server()
            resp = yield from self.client.post(url, data=json.dumps(data))
            self.assertEqual(200, resp.status)
            self.assertEqual(check, (yield from resp.json()))
            self.assertEqual(None, (yield from resp.release()))

        self.loop.run_until_complete(post(INVALID_PARAMS,
                                          create_request("v_hello")))
        self.loop.run_until_complete(
            post(create_response(result={"status": "ok"}),
                 create_request("v_hello", params={"data": "ok"})))
        self.loop.run_until_complete(
            post(create_response(1234, {"status": "ok"}),
                 create_request("v_hello", 1234, {"data": "ok"})))
        self.loop.run_until_complete(
            post(create_response(None, {"status": "ok"}),
                 create_request("v_hello", None, {"data": "ok"})))
        self.loop.run_until_complete(
            post(create_response("1", {"status": "OK"}),
                 create_request("v_hello", "1", {"data": "TEST"})))
        self.loop.run_until_complete(
            post(create_response(True, {"status": "ok"}),
                 create_request("v_hello", True, {"data": "ok"})))
        self.loop.run_until_complete(
            post(create_response(False, {"status": "ok"}),
                 create_request("v_hello", False, {"data": "ok"})))

    def test_without_validate(self):
        @asyncio.coroutine
        def post(check, data=None):
            app, srv, url = yield from self.create_server()
            resp = yield from self.client.post(url, data=json.dumps(data))
            self.assertEqual(200, resp.status)
            self.assertEqual(check, (yield from resp.json()))
            self.assertEqual(None, (yield from resp.release()))

        self.loop.run_until_complete(
            post(create_response(None, {"a": "b"}),
                 create_request("hello")))
        self.loop.run_until_complete(
            post(create_response(123, {"a": "b"}),
                 create_request("hello", 123)))
        self.loop.run_until_complete(
            post(create_response("123", {"a": "b"}),
                 create_request("hello", "123")))


aiohttp_jrpc/__init__.py
""" Simple JSON-RPC 2.0 protocol for aiohttp"""
from .exc import (ParseError, InvalidRequest, InvalidParams,
                  InternalError, InvalidResponse)
from .errors import JError, JResponse

from validictory import validate, ValidationError, SchemaError
from functools import wraps
from uuid import uuid4
from aiohttp import ClientSession
import asyncio
import json
import traceback

__version__ = '0.1.0'

REQ_JSONRPC20 = {
    "type": "object",
    "properties": {
        "jsonrpc": {"pattern": r"2\.0"},
        "method": {"type": "string"},
        "params": {"type": "any"},
        "id": {"type": "any"},
    },
}
RSP_JSONRPC20 = {
    "type": "object",
    "properties": {
        "jsonrpc": {"pattern": r"2\.0"},
        "result": {"type": "any"},
        "id": {"type": "any"},
    },
}
ERR_JSONRPC20 = {
    "type": "object",
    "properties": {
        "jsonrpc": {"pattern": r"2\.0"},
        "error": {
            "type": "object",
            "properties": {
                "code": {"type": "number"},
                "message": {"type": "string"},
            }
        },
        "id": {"type": "any"},
    },
}


@asyncio.coroutine
def jrpc_errorhandler_middleware(app, handler):
    @asyncio.coroutine
    def middleware(request):
        try:
            return (yield from handler(request))
        except Exception:
            traceback.print_exc()
            return JError().internal()
    return middleware


@asyncio.coroutine
def decode(request):
    """ Get/decode/validate json from request """
    try:
        data = yield from request.json(loader=json.loads)
    except Exception as err:
        raise ParseError(err)

    try:
        validate(data, REQ_JSONRPC20)
    except ValidationError as err:
        raise InvalidRequest(err)
    except SchemaError as err:
        raise InternalError(err)
    except Exception as err:
        raise InternalError(err)
    return data


class Service(object):
    """ Service class """

    def __new__(cls, ctx):
        """ Return on call class """
        return cls.__run(cls, ctx)

    def valid(schema=None):
        """ Validation data by specific validictory configuration """
        def dec(fun):
            @wraps(fun)
            def d_func(self, ctx, data, *a, **kw):
                try:
                    validate(data['params'], schema)
                except ValidationError as err:
                    raise InvalidParams(err)
                except SchemaError as err:
                    raise InternalError(err)
                return fun(self, ctx, data['params'], *a, **kw)
            return d_func
        return dec

    @asyncio.coroutine
    def __run(self, ctx):
        """ Run service """
        try:
            data = yield from decode(ctx)
        except ParseError:
            return JError().parse()
        except InvalidRequest:
            return JError().request()
        except InternalError:
            return JError().internal()

        try:
            i_app = getattr(self, data['method'])
            i_app = asyncio.coroutine(i_app)
        except Exception:
            return JError(data).method()

        try:
            resp = yield from i_app(self, ctx, data)
        except InvalidParams:
            return JError(data).params()
        except InternalError:
            return JError(data).internal()

        return JResponse(jsonrpc={
            "id": data['id'], "result": resp
            })


class Response(object):

    __slots__ = ['id', 'error', 'result']

    def __init__(self, id, result=None, error=None, **kw):
        self.id = id
        self.result = result
        self.error = error

    def __repr__(self):
        return "Response(id={rid}, result={res}, error={err}".format(
                    rid=self.id, res=self.result, err=self.error)


class Client(object):
    def __init__(self, url, dumper=None, loop=None):
        self.url = url
        self.dumper = dumper
        if not loop:
            loop = asyncio.get_event_loop()
        if not self.dumper:
            self.dumper = json.dumps

        self.client = ClientSession(
                          loop=loop,
                          headers={'content-type': 'application/json'})

    def __del__(self):
        self.client.close()

    def __encode(self, method, params=None, id=None):
        try:
            data = self.dumper({
                    "jsonrpc": "2.0",
                    "id": id,
                    "method": method,
                    "params": params
                   })
        except Exception as e:
            raise Exception("Can not encode: {}".format(e))

        return data

    @asyncio.coroutine
    def call(self, method, params=None, id=None, schem=None):
        if not id:
            id = uuid4().hex
        try:
            resp = yield from self.client.post(
                   self.url, data=self.__encode(method, params, id))
        except Exception as err:
            raise Exception(err)

        if 200 != resp.status:
            raise InvalidResponse(
                "Error, server retunrned: {status}".format(status=resp.status))

        try:
            data = yield from resp.json()
        except Exception as err:
            raise InvalidResponse(err)

        try:
            validate(data, ERR_JSONRPC20)
            return Response(**data)
        except ValidationError:
            # Passing data to validate response.
            # Good if does not valid to ERR_JSONRPC20 object.
            pass
        except Exception as err:
            raise InvalidResponse(err)

        try:
            validate(data, RSP_JSONRPC20)
            if id != data['id']:
                raise InvalidResponse(
                       "Rsponse id {local} not equal {remote}".format(
                            local=id, remote=data['id']))
        except Exception as err:
            raise InvalidResponse(err)

        if schem:
            try:
                validate(data['result'], schem)
            except ValidationError as err:
                raise InvalidResponse(err)
            except Exception as err:
                raise InternalError(err)

        return Response(**data)


aiohttp_jrpc/errors.py
""" Error responses """
from aiohttp.web import Response
import json


class JResponse(Response):
    """ Modified Reponse from aohttp """
    def __init__(self, *, status=200, reason=None,
                 headers=None, jsonrpc=None):
        if jsonrpc is not None:
            jsonrpc.update({'jsonrpc': '2.0'})
            text = json.dumps(jsonrpc)
        super().__init__(status=status, reason=reason, text=text,
                         headers=headers, content_type='application/json')


class JError(object):
    """ Class with standart errors """
    def __init__(self, data=None, rid=None):
        if data is not None:
            self.rid = data['id']
        else:
            self.rid = rid

    def parse(self):
        """ json parsing error """
        return JResponse(jsonrpc={
            'id': self.rid,
            'error': {'code': -32700, 'message': 'Parse error'},
        })

    def request(self):
        """ incorrect json rpc request """
        return JResponse(jsonrpc={
            'id': self.rid,
            'error': {'code': -32600, 'message': 'Invalid Request'},
        })

    def method(self):
        """ Not found method on the server """
        return JResponse(jsonrpc={
            'id': self.rid,
            'error': {'code': -32601, 'message': 'Method not found'},
        })

    def params(self):
        """ Incorrect params (used in validate) """
        return JResponse(jsonrpc={
            'id': self.rid,
            'error': {'code': -32602, 'message': 'Invalid params'},
        })

    def internal(self):
        """ Internal server error, actually send on every unknow exception """
        return JResponse(jsonrpc={
            'id': self.rid,
            'error': {'code': -32603, 'message': 'Internal error'},
        })

    def custom(self, code, message):
        """
        Specific server side errors use: -32000 to -32099
        reserved for implementation-defined server-errors
        """
        if -32000 < code or -32099 > code:
            code = -32603
            message = 'Internal error'
        return JResponse(jsonrpc={
            'id': self.rid,
            'error': {'code': code, 'message': message},
        })
